### PR TITLE
fix: don't redundantly evaluate the same partitions for compaction

### PR DESCRIPTION
**Pull request by Joe-Blount**
This PR Is a followon and complementary change to https://github.com/influxdata/influxdb_iox/pull/8112 from yesterday changing the same area of code, but for a different issue.

We've seen the compactor spam the catalog with queries when its idle.  But it doesn't do this in every idle circumstance.  If it is completely idle (no partitions seeing any activity) it doesn't hit the catalog very hard.
When its mostly idle, like many partitions all receiving data trickle in, that's when the catalog may get spammed with many requests.

I believe the cause of spamming the catalog in a semi-idle state results from the hard coded 10m lookback for "hot" partitions that may need compacted.  If the compactor is idle enough that it can get through the batch of hot partitions quickly (e.g. 2m), the next query still looks back 10m.  That means most of the partitions its evaluating in the next batch are guaranteed to not need compacted because it just finished evaluating them.  I believe the catalog spam results from this redundant evaluation of partitions that are guaranteed to not need compacted.  Its a potentially large amount of busy-work to keep re-deciding that these partitions don't need compacted.

The PR I did yesterday expands the lookback to catch previously skipped partitions when the query is slow.  This PR builds on that to shrink the lookback when we know we've already evaluated those partitions.

It might be necessary to put a minimum time between the queries for hot partitions.  For example: if we're targetting a 10m definition of "hot" partitions, and capping the lookback at 30m when the partition processing is slow (per yesterday's PR), if the partition processing is fast we could make sure we wait at least 3-5m between getting batches of hot partitions.  This is probably a good thing to do, and it will be another trivial change to this same area of code, but I'm curious to see how this change performs first.

I'd be surprised if this made the idle cluster slam the catalog worse than it already does.  This PR should make the query pattern of the poorly behaved idle cluster more like the query pattern of the well behaved idle clusters.

But this can be reverted if necessary.

helps: https://github.com/influxdata/influxdb_iox/issues/7973

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Review by wiedld — COMMENTED** [review]
Making sure I understand. We have a throttle on the partitions_source, which throttles for any partition which has not had a commit in the past duration_threshold ([set here](https://github.com/influxdata/influxdb_iox/blob/3a8a8a153ebd9c220ed4ff542fbf911d222218f6/compactor/src/components/hardcoded.rs#L161)). Since we made the query time window more dynamic (can be larger) to avoid leaving behind partitions -- should we try increasing this throttle? Or is it better to add the logic in the dynamic time window calculation (as your PR does)?

File: compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs [hunk 1]
```diff
@@ -74,18 +74,21 @@ impl PartitionsSource for CatalogToCompactPartitionsSource {
...
-                    // this might skip something (until cold compaction), but we need a limit in how far we look back.
-                    minimum_time = self.time_provider.now() - self.min_threshold * 3;
-                }
+            // query for partitions with activity since the last query.  We shouldn't query for a time range
+            // we've already covered.  So if the prior query was 2m ago, and the query covered 10m, ending at
+            // the time of that query, we just need to query for activity in the last 2m.  Asking for more than
+            // that creates busy-work that will spam the catalog with more queries to determine no compaction
+            // nneded.  But we also don't want to query so far back in time that we get all partitions, so the
+            // lookback is limited to 3x the configured threshold.
+            if minimum_time < *last || minimum_time.sub(*last) < self.min_threshold * 3 {
```
Comment (wiedld) [inline compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs:83]: Without thinking through the nuances, the conditional flipped from `*last < minimum_time` => to `minimum_time < *last`. With the same output action based upon eval to true. 🤔

File: compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs [hunk 2]
```diff
@@ -122,17 +126,18 @@ mod tests {
         catalog: Arc<MemCatalog>,
         min_threshold: Duration,
         max_threshold: Option<Duration>,
+        second_query_delta: Duration, // time between first and second query
         first_expected_ids: &[i64], // expected values on first fetch, which does a 3x on min_threshold
         second_expected_ids: &[i64], // expected values on second fetch, which uses min_threshold unmodified
     ) {
-        let time_provider = catalog.time_provider();
+        let time_provider = Arc::new(MockProvider::new(catalog.time_provider().now()));
```
Comment (wiedld) [inline compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs:133]: Why do we need this? Because it's import (for other compactor tests which compare time) to use the same time_provider in the catalog & compactor. What am I missing?

Comment (Joe-Blount) [reply compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs:133]: > What am I missing?

Look down 20 lines to this:
`time_provider.inc(second_query_delta);` 

The mock provider lets us set bogus times for testing.  The first and second queries in the tests run microseconds apart, but the time provider is advanced by `second_query_delta` (a minute, or hours, depending on the test case) before the second query.  The tests can simulate queries separated by any arbitrary amount of time, but the test completes in microseconds.

**Comment by Joe-Blount** [comment]
> Making sure I understand. We have a throttle on the partitions_source, which throttles for any partition which has not had a commit in the past duration_threshold ([set here](https://github.com/influxdata/influxdb_iox/blob/3a8a8a153ebd9c220ed4ff542fbf911d222218f6/compactor/src/components/hardcoded.rs#L161)). Since we made the query time window more dynamic (can be larger) to avoid leaving behind partitions -- should we try increasing this throttle? Or is it better to add the logic in the dynamic time window calculation (as your PR does)?

The throttling relevant to this discussion was added [in this PR](https://github.com/influxdata/influxdb_iox/pull/8035) to allow us to slow one of the compactor's queries to the catalog (for partition info, which it must do to evaluate what needs compacted).  I [enabled it yesterday](https://github.com/influxdata/k8s-idpe/pull/25716), and sorta worked, in that it did cap the rate of **those** queries.  But the catalog does other queries, which still were at a high rate.  The query the throttling was added to what looked like a critical path that would allow us to slow the compactor when it doesn't have much work to do.  Maybe it would work if I lowered the setting more - but the rate I set it to was getting low enough I'm concerned it would affect the busy clusters that need to go fast. So although setting it lower might allow it to slow the compactor in this semi-idle scenario, I'm afraid then it would be low enough to inappropriately limit the compactor when it has a lot of work to do.  I'm reluctant to add throttling knobs that require maintenance as a cluster grows.  

We should be able to get that work, and we might need to, but its a less than ideal solution, and I think :crossed_fingers: this PR can make it obsolete...

It was bugging me that the catalog spam situation only occurs on semi-idle systems, combined with the PR I did yesterday, lead me to what I think is the actual culprit of the spam (this hard coded minimum lookback).  My hope is that cleaning up this hardcoded lookback will address the catalog query spam issue, making the throttling unnecessary. Even if it doesn't fix the catalog query spam issue, skipping partitions we know we've already evaluated (and haven't changed since) is a good thing to do anyway.

TLDR: we were working on a bandaide that wasn't working as well as hoped, when I may have stumbled onto the root cause

this is scattered & rambling explanation, but hopefully it makes sense.

File: compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs (hunk 1 above)
Comment (Joe-Blount) [reply compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs:83]: previously it only adjusted the start time when it was necessary to avoid skipping partitions.

now it always adjusts the start time.  It could be that its lengthening it (like before) to avoid skipping.  Or it could be that its shortening it to avoid re-evaluating partitions we just looked at.

File: compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs (hunk 2 above)
Comment (wiedld) [reply compactor_scheduler/src/local_scheduler/partitions_source/catalog_to_compact.rs:133]: I missed that. 😆
